atomkraft/chain/testnet.py

```diff
@@ -311,22 +311,23 @@ def spinup(self):
         for node in self.validator_nodes.values():
             node.start()
         for node in self.validator_nodes.values():
-            while True:
-                try:
-                    addr = node.get(
-                        self.ports()["rpc"].config_file,
-                        self.ports()["rpc"].property_path,
-                    )
-                    ip, port = addr.split("//")[-1].split(":")
-                    with socket.create_connection(
-                        (ip, port),
-                        timeout=5,
-                    ):
-                        break
-                except OSError:
-                    # todo: remove this sleep in future
-                    time.sleep(0.1)
-                    pass
+            for endpoint_type in ["rpc", "grpc"]:
+                while True:
```

Comment:
I don't like this line but it has been working until now, so let's keep it for the moment.

Reply:
Can you explain why you don't like it? I can probably refactor.

Reply:
Breaking the loop depends on the external function `socket.create_connection`, so you never know exactly if the loop would continue forever. I guess it would be better to have a timeout or a maximum number of retries wrapping the loop. But, as I said, it has worked well until now, so don't worry too much.

Reply:
You're right. Maybe we can use something like this https://pypi.org/project/retry2

Reply:
If there's no standard pythonic way of doing it, then maybe this library is an option. I don't like adding external dependencies unless is strictly necessary. Anyway, it's out of scope for this PR, we may consider it for another issue.

Reply:
I added retry support using `tenacity` library. I used 10 minutes timeout on each retry session - which is way too much IMO. When a timeout is hit, it will throw an error. At least this guarantees eventual termination.